Setup sentry

Belco90 opened the pull request:
## Changes

- Set up Sentry for Next.js
- Additionally, Sentry integration for Vercel and GitHub have been set up too

## Context

<!-- If you're fixing an issue with this pull request, use the Fixes keyword with the issue number you're closing, like this:

Fixes #1
-->

Since we are gonna start changing and refactoring stuff, it's good to catch possible errors on production/preview environments.

HonkingGoose commented:
I've never seen or used Sentry. Can you give me a high-level overview of what Sentry can do for us, or will help us with?

Belco90 commented:
Sure! I was planning to add more context but I created the PR as a draft to check Sentry was actually working 😅 

So Sentry is basically a tool to catch and report errors generated at the production platform. You can read more about it [here](https://sentry.io/welcome/).

You can see I've introduced a new button in this PR with a callback that is trying to execute an existing method. When the user clicks that button, nothing will happen in the UI but an error is thrown in JS under the hood. Sentry will catch it and report it to us! This is the error caught by Sentry for that exploding button I introduced for testing here:

![CleanShot 2021-07-12 at 19 41 02@2x](https://user-images.githubusercontent.com/2677072/125332106-19ce4b80-e349-11eb-85b0-2557073ac34a.png)

This is interesting for being aware of run-time errors and detecting when they were introduced. I don't expect this to report too many errors not to be honest, but still is useful, especially if more users start landing to the app from Renovate for example.

Belco90 commented on next-env.d.ts:
```diff
@@ -1,4 +1,3 @@
-/* eslint-disable @typescript-eslint/triple-slash-reference */
```
This is not necessary anymore since `xo` ESLint config was removed.

HonkingGoose replied:
Cool! That line kept getting removed when I ran Prettier, and I wasn't sure if we needed that line or not. Saves me a bit of hassle each commit. 😄

Belco90 commented on next.config.js:
```diff
@@ -14,3 +15,20 @@ module.exports = withBundleAnalyzer({
   },
 })
+
+const SentryWebpackPluginOptions = {
```
The whole config has been automatically generated from `@sentry/wizard` for a Next.js project.

HonkingGoose replied:
We can assume that works then. 😄

Belco90 replied:
We are sure it's working since I forced an issue with that strange button I added right after creating the PR 😅

Belco90 commented on next.config.js:
```diff
@@ -14,3 +15,20 @@ module.exports = withBundleAnalyzer({
+
+  silent: process.env.NODE_ENV !== 'production', // Suppresses all logs in envs other than production
+  deploy: {
+    env: process.env.VERCEL_ENV,
```
This is the only thing I added myself so we can classify errors per environment. [`VERCEL_ENV` is exposed by Vercel](https://vercel.com/docs/environment-variables#system-environment-variables) so this will be either `production`, `preview`, or `development`.

HonkingGoose replied:
That seems handy, you can see which environment is broken. 😄

Belco90 commented on sentry.client.config.js:
```diff
@@ -0,0 +1,17 @@
+
+import * as Sentry from '@sentry/nextjs'
+
+const SENTRY_DSN = process.env.SENTRY_DSN || process.env.NEXT_PUBLIC_SENTRY_DSN
```
I added `NEXT_PUBLIC_SENTRY_DSN` to Vercel env vars.

HonkingGoose reviewed (changes requested):
The build is not passing right now, so maybe you need to fix something?

Belco90 commented:
> The build is not passing right now, so maybe you need to fix something?

Yes, I'm on it. Basically, the command to push some files to Sentry is being executed every time `yarn build` is ran. I'm checking how to opt out of this behavior to have control over it.

HonkingGoose commented on next.config.js:
```diff
@@ -2,8 +2,9 @@
 const withBundleAnalyzer = require('@next/bundle-analyzer')({
   enabled: process.env.ANALYZE === 'true',
 })
+const { withSentryConfig } = require('@sentry/nextjs')
```
```suggestion

const { withSentryConfig } = require('@sentry/nextjs')
```

I'd like a newline here for readability. 😉

Belco90 replied:
I'm sorry but the spacing and sorting for the imports are set up with `eslint-plugin-import`. With our current config, it will group together (i.e. without newlines) all imports belonging to the same category. These two are external imports so they are group together:

<img width="708" alt="CleanShot 2021-07-12 at 20 30 35@2x" src="https://user-images.githubusercontent.com/2677072/125338066-0a9ecc00-e350-11eb-88ce-ad56d9e9fc71.png">

HonkingGoose replied:
Oh well. 😄 

Let's keep things as they are then. 😄

HonkingGoose reviewed (changes requested):
Just one thing, and I think we're good to merge.
